PA-BT Functionality

solarbean23 opened the pull request.

solarbean23 committed:
baseline from simple scenario

solarbean23 committed:
PA-BT Basic Function with Simple DEMO

inmo-jang commented on modules/ppa_bt_constructor.py:
```diff
@@ -0,0 +1,113 @@
+# ppa_bt_expansion.py
+from modules.base_bt_nodes import * # 확인 후 수정
```
It is better to avoid using "*" as this is vague.

solarbean23 committed:
Refactor XML generation logic for PPA-BT

solarbean23 commented:
- Updated XML saving logic to handle pre-conditions more effectively.
- Simplified XML structure by omitting unnecessary Sequence nodes when no pre-conditions are present.
- Improved readability and compatibility with Groot2 by refining indentation and line breaks.
- Fixed fallback and sequence node handling to ensure correct behavior.

inmo-jang commented on modules/ppa_bt_constructor.py:
```diff
@@ -29,7 +32,15 @@ def expand_behavior_tree(tree, failed_condition, ppa_library):
-        save_tree_as_xml(tree, "agent_bt_temp.xml")
+
+        # Set the output directory
+        output_dir = "/home/solarbean23/Desktop/space_dev/space-simulator/scenarios/pa_bt_test/output/xml_output"
```
`output_dir` 경로가 하드코딩 되어 있어서 좋지 않습니다. result saving 하는 부분 (https://github.com/inmo-jang/space-simulator/blob/c45b22428ddab701366d75b190cfe50d9aaef034/modules/utils.py#L104) 참고하셔서 저장될 주소가, 이것과 최대한 동일하게 해주세요. 어떤 것은 저기에 어떤 것은 여기에 저장되면 헷갈릴 것 같습니다.

solarbean23 replied:
"Modify output path" commit에 수정하였습니다!
pa_bt_test 디렉토리의 output폴더에 날짜별로 저장이 되도록 수정하였습니다.
기존에는 활성화된 Decision Making Plugin에 따라 첫 제목이 유동적으로 바뀌었는데, 현재 pa_bt_test를 위해서 따로 plugin을 사용하지 않고 있기 때문에 "PA_BT 날짜".xml로 저장되게끔 수정하였습니다.

solarbean23 committed:
Refactor: Apply consistent CamelCase naming and update task assignment logic

solarbean23 committed:
Modify output path

inmo-jang commented on modules/ppa_bt_constructor.py:
```diff
@@ -1,12 +1,14 @@
 import datetime
 import importlib
 bt_module = importlib.import_module(config.get('scenario').get('environment') + ".bt_nodes")
+result_saver = ResultSaver(config_file_path="/path/to/config.yaml")
```
`config_file_path = "/path/to/config.yaml"` 이 부분 이상한데요? 이런 표현은 "나중에 여기에 이렇게 경로 넣으세요"하는 표현인데, 혹시 이거 코드 실행했을때 문제 없던가요?

solarbean23 replied:
넵 따로 문제는 없었습니다.
현재 제 config.yaml의 saving_options부분에서는 다음과 같이 작성되어 있습니다.
  saving_options:
    output_folder: scenarios/pa_bt_test/output
 그래서 pa_bt_test/output디렉토리에 날짜별로 저장이 되는것을 확인하였습니다. 
![Screenshot from 2025-01-17 09-07-42](https://github.com/user-attachments/assets/21bd1ac0-4095-46ec-b616-2ad0486e87fd)

inmo-jang replied:
흠. 뭔가 좀 이상하게 느껴집니다.  https://github.com/solarbean23/space-simulator/blob/51a8aaa74e75652ceaa4ab5944a0305c570a2ce2/modules/ppa_bt_constructor.py#L39 
```
        file_path = result_saver.generate_output_filename(extension="xml")
```
이 부분에서 결과로 나오는  `file_path` 가 어떻게 출력되는지 한번 체크해서 보여주실래요? 하드코딩으로 `/path/to/...` 라고 되어있는데 동작이 된다는것이 이상하네요.

solarbean23 replied:
![image](https://github.com/user-attachments/assets/670fa81a-434d-4852-84ad-986663e9b3ce)

다음과 같이 출력됩니다.

inmo-jang replied:
지금 보니까 `ResultSaver` 자체에 뭔가 오류가 있습니다. 

일단 당장 오류는 없으니, 
`result_saver = ResultSaver(config_file_path="/path/to/config.yaml")` 에 ` #TODO: Needs debug` 라고 한줄 추가해주시는 걸로 일단 이 건은 마무리 하는 걸로 하시죠.

solarbean23 replied:
넵 수정하겠습니다!

inmo-jang commented on modules/ppa_bt_constructor.py:
```diff
@@ -0,0 +1,197 @@
+        action_class = getattr(bt_module, ppa_fail_entry["action"])
+        action_node = action_class(ppa_fail_entry["action"], None)
+        sequence.children.append(action_node)
+    else:
```
이 부분에서 postcondition중에 precondition도 없고 action도 없는 경우는 에러 발생을 안할까요? (예: IsTargetInfoReady)

solarbean23 replied:
이 부분도 수정이 필요하다고 생각합니다.
다만, 신기한 부분은 Notion에 작성한 것처럼 pre-con과 action이 둘다 없는 post-condition의 위치가 어디에 있는지에 따라 알아서 생성되는 경우도 있고, 오류가 나는 경우도 있었습니다.
이 부분은 디버그하면서 제대로 확인해보겠습니다.

inmo-jang replied:
혹시 이 부분은 현재 어떻게 되었을까요? 디버그 되었나요?

inmo-jang commented on modules/ppa_bt_constructor.py:
```diff
@@ -0,0 +1,197 @@
+        for row in csv_data:
+            Post_condition = row["Post_condition"]
+            Action = row["Action"]
+            Pre_conditions = row["Pre_conditions"].split(",") if row["Pre_conditions"] else []
```
csv file을 csv모듈을 통해 읽는 방식으로 하면 Pre conditions이 2개 이상일 경우에 대해서 처리를 못할 것으로 보임. 
csv file을 그냥 text file파일이라고 가정하고 line-by-line으로 처리하는 방식으로 수정하는 것이 좋을 수도 있음.

inmo-jang commented on agent_bt_temp.xml:
`agent_bt_temp.xml` 파일은 다른 시나리오랑 상관이 없으므로, `scenario/pa_bt_test/` 안으로 넣어주시기 바랍니다.

solarbean23 replied:
agent_bt_temp.xml이 이전에 저장되었던 임시 파일이라, 삭제한 뒤 다시 commit하였습니다!

solarbean23 committed:
Refactor CSV reading & modification, enhance PPA-BT generation

solarbean23 committed:
Delete temp xml file

solarbean23 committed:
Add debug note for ResultSaver initialization in config setup
